## 01. Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

## 02. Import Data Sets

In [2]:
path = r'D:\02.2022_Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data', 'Aftermerge_busiest_dayandtime.pkl'))

In [4]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32399732 entries, 0 to 32435058
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                float64
 1   user_id                 float64
 2   order_number            float64
 3   orders_day_of_week      float64
 4   order_time_of_day       float64
 5   days_since_prior_order  float16
 6   first_order             object 
 7   product_id              float64
 8   add_to_cart_order       float64
 9   reordered               float64
 10  product_name            object 
 11  aisle_id                float64
 12  department_id           float64
 13  prices                  float16
 14  busiest_day             object 
 15  busiest_days            object 
 16  busiest_period_of_day   object 
dtypes: float16(2), float64(10), object(5)
memory usage: 4.0+ GB


## 03. Exercise Walk-through

In [5]:
df = ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 17)

In [7]:
ords_prods_merge.shape

(32399732, 17)

In [8]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   
5  3367565.0      1.0           6.0                 2.0                7.0   
6   550135.0      1.0           7.0                 1.0                9.0   
7  3108588.0      1.0           8.0                 1.0               14.0   
8  2295261.0      1.0           9.0                 1.0               16.0   
9  2550362.0      1.0          10.0                 4.0                8.0   

   days_since_prior_order first_order  product_id  add_to_cart_order  \
0                     NaN        True       196.0                1.0   
1                    15.0       False       196.0                1.0   
2                    21.0       False       196.0                1.0   
3                    29.0       False       196.0                1.0   
4                    28.0       False       196.0                1.0   
5                    19.0       False       196.0                1.0   
6                    20.0       False       196.0                1.0   
7                    14.0       False       196.0                2.0   
8                     0.0       False       196.0                4.0   
9                    30.0       False       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices     busiest_day  \
0        0.0         Soda      77.0            7.0     9.0  Regularly busy   
1        1.0         Soda      77.0            7.0     9.0  Regularly busy   
2        1.0         Soda      77.0            7.0     9.0  Regularly busy   
3        1.0         Soda      77.0            7.0     9.0      Least busy   
4        1.0         Soda      77.0            7.0     9.0      Least busy   
5        1.0         Soda      77.0            7.0     9.0  Regularly busy   
6        1.0         Soda      77.0            7.0     9.0  Regularly busy   
7        1.0         Soda      77.0            7.0     9.0  Regularly busy   
8        1.0         Soda      77.0            7.0     9.0  Regularly busy   
9        1.0         Soda      77.0            7.0     9.0      Least busy   

      busiest_days busiest_period_of_day  
0   Regularly busy        Average orders  
1  Least busy days        Average orders  
2  Least busy days           Most orders  
3  Least busy days        Average orders  
4  Least busy days           Most orders  
5   Regularly busy        Average orders  
6     Busiest days        Average orders  
7     Busiest days           Most orders  
8     Busiest days           Most orders  
9  Least busy days        Average orders

In [9]:
# Grouping data -- invisible
df.groupby('product_name')

In [10]:
# Aggregating data -- single aggregation (1) -better than (3)
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
4.0              18.825780
7.0              17.472355
13.0             17.993423
14.0             19.246334
16.0             19.463012
17.0             11.294069
19.0             19.305237
20.0             17.599636

In [11]:
# Aggregating data -- single aggregation (2) -better than (3)
 # df.groupby('department_id')['order_number'].mean()

In [12]:
# Aggregating data -- single aggregation (3)
# df.groupby('department_id').order_number.mean()

In [13]:
# Aggregating data -- multiple aggregation
# df.groupby('department_id').agg({'order_number':['mean','min','max']})

In [14]:
# Transforming
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\X405U\AppData\Local\Temp/ipykernel_3900/3878628073.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)


In [15]:
# Checking the transformation
df.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329.0      1.0           1.0                 2.0                8.0   
1   2398795.0      1.0           2.0                 3.0                7.0   
2    473747.0      1.0           3.0                 3.0               12.0   
3   2254736.0      1.0           4.0                 4.0                7.0   
4    431534.0      1.0           5.0                 4.0               15.0   
..        ...      ...           ...                 ...                ...   
95  3226575.0    360.0           1.0                 5.0               12.0   
96  1469869.0    377.0           3.0                 5.0               17.0   
97  1927023.0    387.0           2.0                 4.0               10.0   
98   858092.0    420.0           4.0                 1.0               19.0   
99   119314.0    420.0           5.0                 3.0               19.0   

    days_since_prior_order first_order  product_id  add_to_cart_order  \
0                      NaN        True       196.0                1.0   
1                     15.0       False       196.0                1.0   
2                     21.0       False       196.0                1.0   
3                     29.0       False       196.0                1.0   
4                     28.0       False       196.0                1.0   
..                     ...         ...         ...                ...   
95                     NaN        True       196.0                1.0   
96                     3.0       False       196.0                9.0   
97                    22.0       False       196.0                3.0   
98                    30.0       False       196.0                2.0   
99                    16.0       False       196.0                2.0   

    reordered product_name  aisle_id  department_id  prices     busiest_day  \
0         0.0         Soda      77.0            7.0     9.0  Regularly busy   
1         1.0         Soda      77.0            7.0     9.0  Regularly busy   
2         1.0         Soda      77.0            7.0     9.0  Regularly busy   
3         1.0         Soda      77.0            7.0     9.0      Least busy   
4         1.0         Soda      77.0            7.0     9.0      Least busy   
..        ...          ...       ...            ...     ...             ...   
95        0.0         Soda      77.0            7.0     9.0  Regularly busy   
96        0.0         Soda      77.0            7.0     9.0  Regularly busy   
97        0.0         Soda      77.0            7.0     9.0      Least busy   
98        0.0         Soda      77.0            7.0     9.0  Regularly busy   
99        1.0         Soda      77.0            7.0     9.0  Regularly busy   

       busiest_days busiest_period_of_day  max_order  
0    Regularly busy        Average orders       10.0  
1   Least busy days        Average orders       10.0  
2   Least busy days           Most orders       10.0  
3   Least busy days        Average orders       10.0  
4   Least busy days           Most orders       10.0  
..              ...                   ...        ...  
95   Regularly busy           Most orders        1.0  
96   Regularly busy        Average orders        3.0  
97  Least busy days           Most orders        2.0  
98     Busiest days        Average orders        5.0  
99  Least busy days        Average orders        5.0  

[100 rows x 18 columns]

In [16]:
df.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order'],
      dtype='object')

In [17]:
df.shape

(1000000, 18)

In [18]:
pd.options.display.max_rows = None

In [19]:
# Creating a flag
df.loc[df['max_order'] > 40,'loyalty_flag'] = 'Loyal customer'

D:\Anaconda\lib\site-packages\pandas\core\indexing.py:1684: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = infer_fill_value(value)
D:\Anaconda\lib\site-packages\pandas\core\indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [20]:
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10),'loyalty_flag'] = 'Regular customer'

In [21]:
df.loc[df['max_order'] <= 10,'loyalty_flag'] = 'New customer'

In [22]:
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    463844
Loyal customer      335086
New customer        201070
Name: loyalty_flag, dtype: int64

In [23]:
# Checking the flags
df[['user_id','loyalty_flag','order_number']].head(60)

user_id      loyalty_flag  order_number
0       1.0      New customer           1.0
1       1.0      New customer           2.0
2       1.0      New customer           3.0
3       1.0      New customer           4.0
4       1.0      New customer           5.0
5       1.0      New customer           6.0
6       1.0      New customer           7.0
7       1.0      New customer           8.0
8       1.0      New customer           9.0
9       1.0      New customer          10.0
10     15.0  Regular customer          15.0
11     15.0  Regular customer          17.0
12     15.0  Regular customer          18.0
13     15.0  Regular customer          21.0
14     15.0  Regular customer          22.0
15     19.0      New customer           2.0
16     19.0      New customer           5.0
17     19.0      New customer           7.0
18     21.0  Regular customer          10.0
19     31.0  Regular customer          10.0
20     31.0  Regular customer          17.0
21     43.0      New customer           6.0
22     43.0      New customer           9.0
23     52.0  Regular customer           1.0
24     52.0  Regular customer           2.0
25     52.0  Regular customer           3.0
26     52.0  Regular customer           5.0
27     52.0  Regular customer           7.0
28     52.0  Regular customer           9.0
29     52.0  Regular customer          10.0
30     52.0  Regular customer          11.0
31     52.0  Regular customer          12.0
32     52.0  Regular customer          15.0
33     52.0  Regular customer          17.0
34     52.0  Regular customer          18.0
35     52.0  Regular customer          19.0
36     52.0  Regular customer          21.0
37     67.0  Regular customer           1.0
38     67.0  Regular customer           2.0
39     67.0  Regular customer           3.0
40     67.0  Regular customer           4.0
41     67.0  Regular customer           5.0
42     67.0  Regular customer           6.0
43     67.0  Regular customer           8.0
44     67.0  Regular customer           9.0
45     67.0  Regular customer          10.0
46     67.0  Regular customer          11.0
47     67.0  Regular customer          12.0
48     67.0  Regular customer          13.0
49     67.0  Regular customer          14.0
50     67.0  Regular customer          16.0
51     67.0  Regular customer          18.0
52     67.0  Regular customer          19.0
53     67.0  Regular customer          21.0
54     67.0  Regular customer          22.0
55     67.0  Regular customer          23.0
56     81.0      New customer           3.0
57     81.0      New customer           5.0
58     82.0  Regular customer           1.0
59     82.0  Regular customer           3.0

In [24]:
df.shape

(1000000, 19)

## 04. Answer Questions

### (1.) In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [26]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']}) 

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.667853
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

### (2.) Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [28]:
df.groupby('department_id').agg({'order_number': ['mean']}) 

order_number
                      mean
department_id             
4.0              18.825780
7.0              17.472355
13.0             17.993423
14.0             19.246334
16.0             19.463012
17.0             11.294069
19.0             19.305237
20.0             17.599636

The mean values of the whole dataset is larger than the ones in the subset. This makes sense, since we don't have many data in subset as in the whole dataset. The mean value could be biased and the infomation of deprtment_id is incomplete.  

### (3.) Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [30]:
# Group by user_id, aggregate order number, and transform the max order number to max_order column 
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)

In [31]:
# Check the result
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329.0      1.0           1.0                 2.0                8.0   
1   2398795.0      1.0           2.0                 3.0                7.0   
2    473747.0      1.0           3.0                 3.0               12.0   
3   2254736.0      1.0           4.0                 4.0                7.0   
4    431534.0      1.0           5.0                 4.0               15.0   
5   3367565.0      1.0           6.0                 2.0                7.0   
6    550135.0      1.0           7.0                 1.0                9.0   
7   3108588.0      1.0           8.0                 1.0               14.0   
8   2295261.0      1.0           9.0                 1.0               16.0   
9   2550362.0      1.0          10.0                 4.0                8.0   
10  2968173.0     15.0          15.0                 1.0                9.0   
11  1870022.0     15.0          17.0                 2.0               16.0   
12  1911383.0     15.0          18.0                 2.0               11.0   
13  2715276.0     15.0          21.0                 1.0                9.0   
14   487368.0     15.0          22.0                 1.0               10.0   
15  2293453.0     19.0           2.0                 5.0               14.0   
16  1973799.0     19.0           5.0                 6.0               12.0   
17   532817.0     19.0           7.0                 4.0               17.0   
18  1573906.0     21.0          10.0                 3.0               10.0   
19  1593000.0     31.0          10.0                 3.0                8.0   

    days_since_prior_order first_order  product_id  add_to_cart_order  \
0                      NaN        True       196.0                1.0   
1                     15.0       False       196.0                1.0   
2                     21.0       False       196.0                1.0   
3                     29.0       False       196.0                1.0   
4                     28.0       False       196.0                1.0   
5                     19.0       False       196.0                1.0   
6                     20.0       False       196.0                1.0   
7                     14.0       False       196.0                2.0   
8                      0.0       False       196.0                4.0   
9                     30.0       False       196.0                1.0   
10                     7.0       False       196.0                2.0   
11                     8.0       False       196.0                6.0   
12                     7.0       False       196.0                1.0   
13                     7.0       False       196.0                1.0   
14                    14.0       False       196.0                1.0   
15                     6.0       False       196.0                3.0   
16                     8.0       False       196.0               15.0   
17                     6.0       False       196.0                1.0   
18                     6.0       False       196.0                2.0   
19                     7.0       False       196.0               17.0   

    reordered product_name  aisle_id  department_id  prices     busiest_day  \
0         0.0         Soda      77.0            7.0     9.0  Regularly busy   
1         1.0         Soda      77.0            7.0     9.0  Regularly busy   
2         1.0         Soda      77.0            7.0     9.0  Regularly busy   
3         1.0         Soda      77.0            7.0     9.0      Least busy   
4         1.0         Soda      77.0            7.0     9.0      Least busy   
5         1.0         Soda      77.0            7.0     9.0  Regularly busy   
6         1.0         Soda      77.0            7.0     9.0  Regularly busy   
7         1.0         Soda      77.0            7.0     9.0  Regularly busy   
8         1.0         Soda      77.0            7.0     9.0  Regularly busy   
9         1.0        

In [32]:
ords_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order'],
      dtype='object')

In [33]:
ords_prods_merge.shape

(32399732, 18)

In [34]:
# Create loyalty flags by loc functions
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [35]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [36]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [37]:
# Check the loyalty flags 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15874128
Loyal customer      10282763
New customer         6242841
Name: loyalty_flag, dtype: int64

In [38]:
ords_prods_merge[['user_id','loyalty_flag','order_number']].head(30)

user_id      loyalty_flag  order_number
0       1.0      New customer           1.0
1       1.0      New customer           2.0
2       1.0      New customer           3.0
3       1.0      New customer           4.0
4       1.0      New customer           5.0
5       1.0      New customer           6.0
6       1.0      New customer           7.0
7       1.0      New customer           8.0
8       1.0      New customer           9.0
9       1.0      New customer          10.0
10     15.0  Regular customer          15.0
11     15.0  Regular customer          17.0
12     15.0  Regular customer          18.0
13     15.0  Regular customer          21.0
14     15.0  Regular customer          22.0
15     19.0      New customer           2.0
16     19.0      New customer           5.0
17     19.0      New customer           7.0
18     21.0  Regular customer          10.0
19     31.0  Regular customer          10.0
20     31.0  Regular customer          17.0
21     43.0  Regular customer           6.0
22     43.0  Regular customer           9.0
23     52.0  Regular customer           1.0
24     52.0  Regular customer           2.0
25     52.0  Regular customer           3.0
26     52.0  Regular customer           5.0
27     52.0  Regular customer           7.0
28     52.0  Regular customer           9.0
29     52.0  Regular customer          10.0

In [39]:
ords_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag'],
      dtype='object')

In [40]:
ords_prods_merge.shape

(32399732, 19)

### (4.) The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [42]:
ords_prods_merge.groupby('loyalty_flag')['prices'].mean() 

loyalty_flag
Loyal customer      7.773438
New customer        7.800781
Regular customer    7.796875
Name: prices, dtype: float16

There is no big difference of the product prices for each loyalty category.

### (5.) The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [44]:
# Descriptive Analyses
ords_prods_merge.groupby('user_id').agg({'prices' : ['mean','min','max']}).head(15)

prices                     
             mean       min        max
user_id                               
1.0      6.367188  1.000000  14.000000
2.0      7.515625  1.299805  14.796875
3.0      8.195312  1.299805  14.398438
4.0      8.203125  1.400391  14.601562
5.0      9.187500  3.199219  14.796875
6.0      8.468750  1.799805  19.593750
7.0      7.464844  1.099609  23.203125
8.0      8.781250  1.099609  14.398438
9.0      7.609375  1.200195  19.296875
10.0     8.648438  1.000000  24.203125
11.0     8.328125  1.400391  14.703125
12.0     8.117188  1.099609  14.601562
13.0     8.710938  1.000000  13.601562
14.0     8.015625  1.099609  23.000000
15.0     3.980469  1.900391   9.000000

In [45]:
# Group by user_id, aggregate prices and trandform the mean price to the spending column
ords_prods_merge['spending'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [47]:
# Create spending flags by loc functions
ords_prods_merge.loc[ords_prods_merge['spending'] < 10, 'spending_flag'] = 'Low spender'

In [48]:
ords_prods_merge.loc[ords_prods_merge['spending'] >= 10, 'spending_flag'] = 'High spender'

In [49]:
# Check the result of loc functions
ords_prods_merge[['user_id','order_number','prices','spending','spending_flag']].head(20)

user_id  order_number  prices  spending spending_flag
0       1.0           1.0     9.0  6.367188   Low spender
1       1.0           2.0     9.0  6.367188   Low spender
2       1.0           3.0     9.0  6.367188   Low spender
3       1.0           4.0     9.0  6.367188   Low spender
4       1.0           5.0     9.0  6.367188   Low spender
5       1.0           6.0     9.0  6.367188   Low spender
6       1.0           7.0     9.0  6.367188   Low spender
7       1.0           8.0     9.0  6.367188   Low spender
8       1.0           9.0     9.0  6.367188   Low spender
9       1.0          10.0     9.0  6.367188   Low spender
10     15.0          15.0     9.0  3.980469   Low spender
11     15.0          17.0     9.0  3.980469   Low spender
12     15.0          18.0     9.0  3.980469   Low spender
13     15.0          21.0     9.0  3.980469   Low spender
14     15.0          22.0     9.0  3.980469   Low spender
15     19.0           2.0     9.0  8.375000   Low spender
16     19.0           5.0     9.0  8.375000   Low spender
17     19.0           7.0     9.0  8.375000   Low spender
18     21.0          10.0     9.0  7.011719   Low spender
19     31.0          10.0     9.0  8.140625   Low spender

In [50]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     32279618
High spender      120114
Name: spending_flag, dtype: int64

In [51]:
ords_prods_merge.shape

(32399732, 21)

### (6.) In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [53]:
# Group by user_id and aggreagte days_since_prior_order in median value
ords_prods_merge.groupby('user_id').agg({'days_since_prior_order': ['median']}).head(5)

days_since_prior_order
                        median
user_id                       
1.0                       20.5
2.0                       13.0
3.0                       10.0
4.0                       20.0
5.0                       11.0

In [54]:
# Group by user_id, aggreagte days_since_prior_order in median value, and transform the value to frequency column
ords_prods_merge['frequency'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(np.median)

In [56]:
# Create frequency flags by loc functions
ords_prods_merge.loc[ords_prods_merge['frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [57]:
ords_prods_merge.loc[(ords_prods_merge['frequency'] > 10) & (ords_prods_merge['frequency'] <= 20), 'frequency_flag'] = 'Regular customer'

In [58]:
ords_prods_merge.loc[ords_prods_merge['frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [59]:
# Check the result for loc functions
ords_prods_merge[['user_id','frequency','frequency_flag']].head(10)

user_id  frequency         frequency_flag
0      1.0       20.5  Non-frequent customer
1      1.0       20.5  Non-frequent customer
2      1.0       20.5  Non-frequent customer
3      1.0       20.5  Non-frequent customer
4      1.0       20.5  Non-frequent customer
5      1.0       20.5  Non-frequent customer
6      1.0       20.5  Non-frequent customer
7      1.0       20.5  Non-frequent customer
8      1.0       20.5  Non-frequent customer
9      1.0       20.5  Non-frequent customer

In [60]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        21556644
Regular customer          7207340
Non-frequent customer     3635743
NaN                             5
Name: frequency_flag, dtype: int64

In [61]:
ords_prods_merge[ords_prods_merge['frequency_flag'].isnull() == True]

order_id   user_id  order_number  orders_day_of_week  \
13654206  895835.0  159838.0           1.0                 0.0   
17278260  895835.0  159838.0           1.0                 0.0   
17649037  895835.0  159838.0           1.0                 0.0   
24164863  895835.0  159838.0           1.0                 0.0   
25907961  895835.0  159838.0           1.0                 0.0   

          order_time_of_day  days_since_prior_order first_order  product_id  \
13654206               17.0                     NaN        True     10749.0   
17278260               17.0                     NaN        True     33401.0   
17649037               17.0                     NaN        True     23695.0   
24164863               17.0                     NaN        True     21334.0   
25907961               17.0                     NaN        True     22198.0   

          add_to_cart_order  reordered  ...     prices  busiest_day  \
13654206                3.0        0.0  ...   5.699219  Busiest day   
17278260                6.0        0.0  ...  14.796875  Busiest day   
17649037                2.0        0.0  ...   4.699219  Busiest day   
24164863                5.0        0.0  ...  10.203125  Busiest day   
25907961                1.0        0.0  ...   1.700195  Busiest day   

          busiest_days  busiest_period_of_day max_order  loyalty_flag  \
13654206  Busiest days         Average orders       1.0  New customer   
17278260  Busiest days         Average orders       1.0  New customer   
17649037  Busiest days         Average orders       1.0  New customer   
24164863  Busiest days         Average orders       1.0  New customer   
25907961  Busiest days         Average orders       1.0  New customer   

          spending  spending_flag frequency  frequency_flag  
13654206  7.417969    Low spender       NaN             NaN  
17278260  7.417969    Low spender       NaN             NaN  
17649037  7.417969    Low spender       NaN             NaN  
24164863  7.417969    Low spender       NaN             NaN  
25907961  7.417969    Low spender       NaN             NaN  

[5 rows x 23 columns]

Missing values come from user_id 159838, who don't have previous order. We can filter out these rows since we just want to categorize customers in three categories based on the frequency. 

In [62]:
ords_prods_merge_noNaN=  ords_prods_merge[ords_prods_merge['frequency_flag'].isnull() == False]

In [63]:
# Exporting
ords_prods_merge_noNaN.to_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_agg.pkl'))

In [64]:
ords_prods_merge_noNaN.to_csv(os.path.join(path,'02 Data','Prepared Data','ords_prods_agg.csv'))